client.go

```diff
@@ -178,6 +242,28 @@ func (c *Client) ExchangeContext(ctx context.Context, m *Msg, a string) (
 	return r, rtt, nil
 }
 
+// Client.DialTimeout acts like DialTimeout, but uses the client's configuration
+func (c *Client) DialTimeoutFunc(network, address string) (conn *Conn, err error) {
+	dialer := net.Dialer{Timeout: c.dialTimeout(), LocalAddr: c.LocalAddr}
+
+	conn = new(Conn)
+	if conn.Conn, err = dialer.Dial(network, address); err != nil {
+		return nil, err
+	}
+	return conn, nil
+}
+
+// Client.DialTimeoutWithTLS acts like DialTimeoutWithTLS, but uses the client's configuration
+func (c *Client) DialTimeoutWithTLS(network, address string) (conn *Conn, err error) {
+	dialer := net.Dialer{Timeout: c.dialTimeout(), LocalAddr: c.LocalAddr}
+
+	conn = new(Conn)
+	if conn.Conn, err = tls.DialWithDialer(&dialer, network, address, c.TLSConfig); err != nil {
```

Comment:
If I understand this correctly, this is similar to what I suggested at https://github.com/miekg/dns/issues/501#issuecomment-320787105 (point 2). I am in favour of a new API that encompasses a `net.Dialer`, while keeping the old API alive but deprecated (until it gets removed).

For this specific change, I'd keep it scoped, so the refactoring could be done just after.